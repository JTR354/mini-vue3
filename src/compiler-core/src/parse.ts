import { NodeTypes } from "./ast";

enum TagTypes {
  START,
  END,
}

export function baseParse(content: string) {
  const context = createContext(content);
  return createRoot(parseChildren(context, []));
}

function parseChildren(context, ancestor) {
  const result: any[] = [];
  while (!isEnd(context)) {
    const s = context.source;
    if (s.startsWith("{{")) {
      result.push(parseInterpolation(context));
    } else if (s.startsWith("<")) {
      result.push(parseElement(context, ancestor));
    } else {
      result.push(parseText(context));
    }
  }

  return result;
}

function isEnd(context) {
  const s = context.source;
  if (s.startsWith("</")) {
    return true;
  }
  return !s;
}

function parseInterpolation(context) {
  const openDelimiter = "{{";
  const closeDelimiter = "}}";
  const source = context.source;
  advanceBy(context, openDelimiter.length);
  const closeIndex = source.indexOf(closeDelimiter);
  const rawContentLength = closeIndex - closeDelimiter.length;
  const rawContent = parseTextData(context, rawContentLength);
  // console.log(context, parseInterpolation.name, 1);
  advanceBy(context, closeDelimiter.length);
  // console.log(context, parseInterpolation.name, 2);
  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content: rawContent.trim(),
    },
  };
}

function advanceBy(context: any, index: number) {
  context.source = context.source.slice(index);
}

function createRoot(children: any[]) {
  return {
    children,
  };
}

function createContext(content: string) {
  return {
    source: content,
  };
}
function parseElement(context: any, ancestor): any {
  const element: any = parseTag(context, TagTypes.START);
  ancestor.push(element.tag);
  element.children = parseChildren(context, ancestor);
  const tag = parseTag(context, TagTypes.END);
  if (ancestor.pop() !== tag) {
    throw new Error(`缺少标签${"span"}`);
  }
  return element;
}
function parseTag(context: any, type: TagTypes) {
  const match = context.source.match(/^<\/?([a-z]*)/i);
  // console.log(match, "parseTag", context);
  const tag = match[1];
  advanceBy(context, match[0].length + 1);
  // console.log(context);
  if (type === TagTypes.END) return tag;
  return {
    type: NodeTypes.ELEMENT,
    tag,
    children: [],
  };
}
function parseText(context: any): any {
  const endToken = ["</", "{{"];
  let endIndex = context.source.length;

  for (const token of endToken) {
    const index = context.source.indexOf(token);
    if (index !== -1 && index < endIndex) {
      endIndex = index;
    }
  }

  const content = parseTextData(context, endIndex);
  // console.log(context, "parse text");
  return {
    type: NodeTypes.TEXT,
    content,
  };
}

function parseTextData(context, length: number) {
  const content = context.source.slice(0, length);
  advanceBy(context, length);
  return content;
}
