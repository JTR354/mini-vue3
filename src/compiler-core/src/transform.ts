export function transform(root, options = {}) {
  const context = createTransformContext(root, options);
  traverseNode(root, context);

  createRootCodegen(root);
}

function createRootCodegen(root: any) {
  root.codegenNode = root.children[0];
}

function traverseNode(node, context) {
  const { nodeTransform } = context;
  for (const fn of nodeTransform) {
    fn(node);
  }
  const children = node.children;

  if (children) {
    for (const child of children) {
      traverseNode(child, context);
    }
  }
}

function createTransformContext(root, options) {
  return {
    root,
    nodeTransform: options.nodeTransform || [],
  };
}
