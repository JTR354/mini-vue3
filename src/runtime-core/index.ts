// export const h = (type, props?, children?) => {
//   return createVNode(type, props, children);
// };

// function createVNode(type, props, children) {
//   return {
//     type,
//     props,
//     children,
//   };
// }

// export const createApp = (rootComponent) => {
//   return {
//     mount(container) {
//       const context = rootComponent.setup();
//       const vnode = rootComponent.render.call(context);
//       const { type, props, children } = vnode;
//       const el: HTMLElement = document.createElement(type);
//       el.textContent = children;
//       for (let key in props) {
//         let value = props[key];
//         if (Array.isArray(value)) {
//           value = value.join(" ");
//         }
//         el.setAttribute(key, value);
//       }
//       const root: HTMLElement = document.querySelector(container);
//       root.append(el);
//     },
//   };
// };

// export { createApp } from "./creatApp";
// export { h } from "./h";
export * from "./h";
// export * from "./createApp";
export { renderSlots } from "./helpers/renderSlots";
export { createTextVNode, createElementVNode } from "./vnode";
export { getCurrentInstance, registerRuntimeCompiler } from "./component";
export { inject, provide } from "./apiInject";
export { createRenderer } from "./renderer";
export { nextTick } from "./scheduler";
export { toDisplayString } from "../shared";
