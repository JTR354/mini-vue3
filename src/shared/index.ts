export * from "./toDisplayString";
export * from "./ShapeFlags";

export const extend = Object.assign;

export const isObject = (value) => {
  return value !== null && typeof value === "object";
};

export const hasChanged = (newValue, value) => {
  return !Object.is(newValue, value);
};

export const hasOwn = (obj: object, key: string) =>
  obj && Object.prototype.hasOwnProperty.call(obj, key);

export const capitalize = (str: string) =>
  str.charAt(0).toUpperCase() + str.slice(1);

export const camelias = (str: string) =>
  str.replace(/-(\w)/g, (_, c: string) => (c ? c.toUpperCase() : ""));

export const toHandlerKey = (event: string) => "on" + capitalize(event);

export const isOn = (key) => /^on[A-Z]/.test(key);

export const EMPTY_OBJ = {};

export const isString = (val) => typeof val === "string";
