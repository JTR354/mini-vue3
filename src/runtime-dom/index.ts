import { createRenderer } from "../runtime-core";
import { isOn } from "../shared";

function createElement(type) {
  return document.createElement(type);
}

function pathProp(key, value, el) {
  if (key === "class" && Array.isArray(value)) {
    value = value.join(" ");
  }
  if (isOn(key)) {
    const event = key.slice(2).toLocaleLowerCase();
    el.addEventListener(event, value);
  } else {
    el.setAttribute(key, value);
  }
}

function insert(el, parent) {
  parent.append(el);
}

function setElementText(el, text) {
  el.textContent = text;
}

const options = {
  createElement,
  pathProp,
  insert,
  setElementText,
};

const renderer = createRenderer(options);

export const createApp = renderer.createApp;

export * from "../runtime-core";
