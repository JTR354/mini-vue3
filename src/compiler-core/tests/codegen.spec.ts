import { describe, it, expect, test } from "vitest";
import { generate } from "../src/codegen";
import { transform } from "../src/transform";
import { baseParse } from "../src/parse";
import { transformElement } from "../src/transforms/transformElement";
import { transformExpression } from "../src/transforms/transformExpression";
import { transformText } from "../src/transforms/transformText";

describe("codegen", () => {
  it("string", () => {
    let ast: any = baseParse("hi");
    transform(ast);
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });

  it("interpolation module", () => {
    const ast = baseParse("{{hello}}");
    transform(ast, {
      nodeTransforms: [transformExpression],
    });

    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });

  it("element and interpolation", () => {
    const ast = baseParse("<div>hi,{{msg}}</div>");
    transform(ast, {
      nodeTransforms: [transformElement, transformText, transformExpression],
    });

    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });
});
