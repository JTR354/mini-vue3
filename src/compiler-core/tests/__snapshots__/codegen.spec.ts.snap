// Vitest Snapshot v1

exports[`codegen > element and interpolation 1`] = `
"
        const { toDisplayString : _toDisplayString, createElementVNode : _createElementVNode} = Vue 

      
return function render(_ctx) {return _createElementVNode('div', null, 'hi,' + _toDisplayString(_ctx.msg))}"
`;

exports[`codegen > interpolation module 1`] = `
"
        const { toDisplayString : _toDisplayString} = Vue 

      
return function render(_ctx) {return _toDisplayString(_ctx.hello)}"
`;

exports[`codegen > string 1`] = `
"
return function render(_ctx) {return 'hi'}"
`;
